diff -ruN binutils-2.21.1-orig/bfd/config.bfd binutils-2.21.1/bfd/config.bfd
--- binutils-2.21.1-orig/bfd/config.bfd	2011-08-21 15:29:37.078052200 +0200
+++ binutils-2.21.1/bfd/config.bfd	2011-08-21 15:49:10.684178600 +0200
@@ -647,6 +647,11 @@
     want64=true
     targ_underscore=no
     ;;
+  x86_64-*-hitomi*)
+    targ_defvec=bfd_elf64_x86_64_vec
+    targ_selvecs="bfd_elf32_i386_vec x86_64pei_vec bfd_elf64_l1om_vec i386coff_vec"
+    want64=true
+    ;;
 #endif
   i[3-7]86-*-lynxos*)
     targ_defvec=bfd_elf32_i386_vec
diff -ruN binutils-2.21.1-orig/config.sub binutils-2.21.1/config.sub
--- binutils-2.21.1-orig/config.sub	2011-08-21 15:29:42.314351700 +0200
+++ binutils-2.21.1/config.sub	2011-08-21 15:41:20.536287700 +0200
@@ -1297,7 +1297,7 @@
 	      | -hpux* | -unos* | -osf* | -luna* | -dgux* | -auroraux* | -solaris* \
 	      | -sym* | -kopensolaris* \
 	      | -amigaos* | -amigados* | -msdos* | -newsos* | -unicos* | -aof* \
-	      | -aos* | -aros* \
+	      | -aos* | -aros* | -hitomi* \
 	      | -nindy* | -vxsim* | -vxworks* | -ebmon* | -hms* | -mvs* \
 	      | -clix* | -riscos* | -uniplus* | -iris* | -rtu* | -xenix* \
 	      | -hiux* | -386bsd* | -knetbsd* | -mirbsd* | -netbsd* \
diff -ruN binutils-2.21.1-orig/gas/configure.tgt binutils-2.21.1/gas/configure.tgt
--- binutils-2.21.1-orig/gas/configure.tgt	2011-08-21 15:29:46.823609600 +0200
+++ binutils-2.21.1/gas/configure.tgt	2011-08-21 15:54:40.008014900 +0200
@@ -179,6 +179,7 @@
   i386-*-beospe*)			fmt=coff em=pe ;;
   i386-*-beos*)				fmt=elf ;;
   i386-*-coff)				fmt=coff ;;
+  i386-*-hitomi*)			fmt=elf ;;
   i386-*-elf)				fmt=elf ;;
   i386-*-kaos*)				fmt=elf ;;
   i386-*-bsd*)				fmt=aout em=386bsd ;;
diff -ruN binutils-2.21.1-orig/ld/Makefile.in binutils-2.21.1/ld/Makefile.in
--- binutils-2.21.1-orig/ld/Makefile.in	2011-08-21 15:30:34.344327600 +0200
+++ binutils-2.21.1/ld/Makefile.in	2011-08-21 16:13:03.923155200 +0200
@@ -766,6 +766,7 @@
 	eelf_x86_64.c \
 	eelf_x86_64_fbsd.c \
 	eelf_x86_64_sol2.c \
+	ehitomi_x88_64.c \
 	ehppa64linux.c \
 	emmo.c \
 	eshelf64.c \
@@ -1154,6 +1155,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_x86_64.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_x86_64_fbsd.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_x86_64_sol2.Po@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/ehitomi_x86_64.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/egld960.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/egld960coff.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eh8300.Po@am__quote@
@@ -3316,6 +3318,9 @@
   $(srcdir)/emultempl/solaris2.em \
   $(ELF_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
 	${GENSCRIPTS} elf_x86_64_sol2 "$(tdir_elf_x86_64_sol2)"
+ehitomi_x86_64.c: $(srcdir)/emulparams/hitomi_x86_64.sh \
+  $(ELF_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
+	${GENSCRIPTS} hitomi_x86_64 "$(tdir_hitomi_x86_64)"
 ehppa64linux.c: $(srcdir)/emulparams/hppa64linux.sh \
   $(ELF_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
 	${GENSCRIPTS} hppa64linux "$(tdir_hppa64linux)"
diff -ruN binutils-2.21.1-orig/ld/configure.tgt binutils-2.21.1/ld/configure.tgt
--- binutils-2.21.1-orig/ld/configure.tgt	2011-08-21 15:30:32.552225100 +0200
+++ binutils-2.21.1/ld/configure.tgt	2011-08-21 15:58:49.981312500 +0200
@@ -207,6 +207,9 @@
 			targ_extra_libpath=elf_i386
 			tdir_i386linux=`echo ${targ_alias}aout | sed -e 's/x86_64/i386/'`
 			tdir_elf_i386=`echo ${targ_alias} | sed -e 's/x86_64/i386/'` ;;
+x86_64-*-hitomi*)	targ_emul=hitomi_x86_64
+			targ_extra_emuls="elf_i386 elf_l1om"
+			;;
 i[3-7]86-*-sysv[45]*)	targ_emul=elf_i386 ;;
 i[3-7]86-*-solaris2*)	targ_emul=elf_i386_sol2
                         targ_extra_emuls="elf_i386_ldso elf_i386 elf_x86_64_sol2 elf_x86_64 elf_l1om"
diff -ruN binutils-2.21.1-orig/ld/emulparams/hitomi_x86_64.sh binutils-2.21.1/ld/emulparams/hitomi_x86_64.sh
--- binutils-2.21.1-orig/ld/emulparams/hitomi_x86_64.sh	1970-01-01 01:00:00.000000000 +0100
+++ binutils-2.21.1/ld/emulparams/hitomi_x86_64.sh	2011-08-21 16:07:58.860706600 +0200
@@ -0,0 +1,17 @@
+SCRIPT_NAME=elf
+ELFSIZE=64
+OUTPUT_FORMAT="elf64-x86-64"
+NO_REL_RELOCS=yes
+TEXT_START_ADDR=0x400000
+MAXPAGESIZE="CONSTANT (MAXPAGESIZE)"
+COMMONPAGESIZE="CONSTANT (COMMONPAGESIZE)"
+ARCH="i386:x86-64"
+MACHINE=
+NOP=0x90909090
+TEMPLATE_NAME=elf32
+GENERATE_SHLIB_SCRIPT=yes
+GENERATE_PIE_SCRIPT=yes
+NO_SMALL_DATA=yes
+LARGE_SECTIONS=yes
+SEPARATE_GOTPLT=24
+IREL_IN_PLT=
