diff -ruN gdc-orig/d/dmd/mangle.c gdc/d/dmd/mangle.c
--- gdc-orig/d/dmd/mangle.c	2011-09-16 07:42:18.824516716 +0200
+++ gdc/d/dmd/mangle.c	2011-09-10 02:30:08.432486090 +0200
@@ -30,7 +30,7 @@
 #include "id.h"
 #include "module.h"
 
-#if TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS
+#if TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_HITOMI
 char *cpp_mangle(Dsymbol *s);
 #endif
 
@@ -125,7 +125,7 @@
                     return ident->toChars();
 
                 case LINKcpp:
-#if DMDV2 && (TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS)
+#if DMDV2 && (TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_HITOMI)
                     return cpp_mangle(this);
 #else
                     // Windows C++ mangling is done by C++ back end
diff -ruN gdc-orig/d/dmd/mars.c gdc/d/dmd/mars.c
--- gdc-orig/d/dmd/mars.c	2011-09-16 07:42:18.824516716 +0200
+++ gdc/d/dmd/mars.c	2011-09-16 07:35:58.404511046 +0200
@@ -66,7 +66,7 @@
 #ifndef IN_GCC
 #if TARGET_WINDOS
     obj_ext  = "obj";
-#elif TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS
+#elif TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_HITOMI
     obj_ext  = "o";
 #elif TARGET_NET
 #else
@@ -79,7 +79,7 @@
 #ifndef IN_GCC
 #if TARGET_WINDOS
     lib_ext  = "lib";
-#elif TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS
+#elif TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_HITOMI
     lib_ext  = "a";
 #elif TARGET_NET
 #else
@@ -92,7 +92,7 @@
 #ifndef IN_GCC
 #if TARGET_WINDOS
     dll_ext  = "dll";
-#elif TARGET_LINUX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS
+#elif TARGET_LINUX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_HITOMI
     dll_ext  = "so";
 #elif TARGET_OSX
     dll_ext = "dylib";
@@ -389,7 +389,7 @@
 
 #if TARGET_WINDOS
     global.params.defaultlibname = "phobos";
-#elif TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS
+#elif TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_HITOMI
     global.params.defaultlibname = "phobos";
 #elif TARGET_NET
 #else
@@ -429,6 +429,10 @@
     VersionCondition::addPredefinedGlobalIdent("Posix");
     VersionCondition::addPredefinedGlobalIdent("Solaris");
     global.params.isSolaris = 1;
+#elif TARGET_HITOMI
+    VersionCondition::addPredefinedGlobalIdent("Posix");
+    VersionCondition::addPredefinedGlobalIdent("Hitomi");
+    global.params.isHitomi = 1;
 #else
 #error "fix this"
 #endif
@@ -467,7 +471,7 @@
                 global.params.link = 0;
             else if (strcmp(p + 1, "cov") == 0)
                 global.params.cov = 1;
-#if TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS
+#if TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_HITOMI
             else if (strcmp(p + 1, "fPIC") == 0)
                 global.params.pic = 1;
 #endif
diff -ruN gdc-orig/d/dmd/mars.h gdc/d/dmd/mars.h
--- gdc-orig/d/dmd/mars.h	2011-09-16 07:42:18.824516716 +0200
+++ gdc/d/dmd/mars.h	2011-09-16 07:35:58.404511046 +0200
@@ -57,6 +57,7 @@
         TARGET_FREEBSD  Covers 32 and 64 bit FreeBSD
         TARGET_OPENBSD  Covers 32 and 64 bit OpenBSD
         TARGET_SOLARIS  Covers 32 and 64 bit Solaris
+        TARGET_HITOMI   Covers 32 and 64 bit Hitomi
         TARGET_NET      Covers .Net
 
     It is expected that the compiler for each platform will be able
@@ -105,7 +106,7 @@
 #define MODULEINFO_IS_STRUCT DMDV2   // if ModuleInfo is a struct rather than a class
 
 // Set if C++ mangling is done by the front end
-#define CPP_MANGLE (DMDV2 && (TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS))
+#define CPP_MANGLE (DMDV2 && (TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_HITOMI))
 
 /* Other targets are TARGET_LINUX, TARGET_OSX, TARGET_FREEBSD, TARGET_OPENBSD and
  * TARGET_SOLARIS, which are
@@ -121,7 +122,7 @@
 #endif
 #endif
 
-#if TARGET_LINUX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS
+#if TARGET_LINUX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_HITOMI
 #ifndef ELFOBJ
 #define ELFOBJ 1
 #endif
@@ -166,6 +167,7 @@
     char isFreeBSD;     // generate code for FreeBSD
     char isOPenBSD;     // generate code for OpenBSD
     char isSolaris;     // generate code for Solaris
+    char isHitomi;      // generate code for Hitomi
     char scheduler;     // which scheduler to use
     char useDeprecated; // allow use of deprecated features
     char useAssert;     // generate runtime code for assert()'s
diff -ruN gdc-orig/d/dmd/mtype.c gdc/d/dmd/mtype.c
--- gdc-orig/d/dmd/mtype.c	2011-09-16 07:42:18.834516715 +0200
+++ gdc/d/dmd/mtype.c	2011-09-16 07:35:58.434511041 +0200
@@ -86,7 +86,7 @@
 int REALSIZE = 16;
 int REALPAD = 6;
 int REALALIGNSIZE = 16;
-#elif TARGET_LINUX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_UNIX
+#elif TARGET_LINUX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_UNIX || TARGET_HITOMI
 int REALSIZE = 12;
 int REALPAD = 2;
 int REALALIGNSIZE = 4;
@@ -253,7 +253,7 @@
     if (global.params.is64bit)
     {
         PTRSIZE = 8;
-        if (global.params.isLinux || global.params.isFreeBSD || global.params.isSolaris)
+        if (global.params.isLinux || global.params.isFreeBSD || global.params.isSolaris || global.params.isHitomi)
         {
             REALSIZE = 16;
             REALPAD = 6;
@@ -268,7 +268,7 @@
 #if TARGET_OSX
         REALSIZE = 16;
         REALPAD = 6;
-#elif TARGET_LINUX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_UNIX
+#elif TARGET_LINUX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_UNIX || TARGET_HITOMI
         REALSIZE = 12;
         REALPAD = 2;
 #else
@@ -1101,7 +1101,7 @@
             sz = REALALIGNSIZE;
             break;
 
-#if TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_UNIX
+#if TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_UNIX || TARGET_HITOMI
         case Tint64:
         case Tuns64:
             sz = global.params.is64bit ? 8 : 4;
diff -ruN gdc-orig/d/dmd2/mangle.c gdc/d/dmd2/mangle.c
--- gdc-orig/d/dmd2/mangle.c	2011-09-16 07:42:18.924516713 +0200
+++ gdc/d/dmd2/mangle.c	2011-09-10 02:30:08.442486090 +0200
@@ -30,7 +30,7 @@
 #include "id.h"
 #include "module.h"
 
-#if IN_GCC || TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS
+#if IN_GCC || TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_HITOMI
 char *cpp_mangle(Dsymbol *s);
 #endif
 
diff -ruN gdc-orig/d/dmd2/mars.c gdc/d/dmd2/mars.c
--- gdc-orig/d/dmd2/mars.c	2011-09-16 07:42:18.924516713 +0200
+++ gdc/d/dmd2/mars.c	2011-09-16 07:35:58.544511046 +0200
@@ -71,7 +71,7 @@
 #ifndef IN_GCC
 #if TARGET_WINDOS
     obj_ext  = "obj";
-#elif TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS
+#elif TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_HITOMI
     obj_ext  = "o";
 #elif TARGET_NET
 #else
@@ -84,7 +84,7 @@
 #ifndef IN_GCC
 #if TARGET_WINDOS
     lib_ext  = "lib";
-#elif TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS
+#elif TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_HITOMI
     lib_ext  = "a";
 #elif TARGET_NET
 #else
@@ -97,7 +97,7 @@
 #ifndef IN_GCC
 #if TARGET_WINDOS
     dll_ext  = "dll";
-#elif TARGET_LINUX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS
+#elif TARGET_LINUX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_HITOMI
     dll_ext  = "so";
 #elif TARGET_OSX
     dll_ext = "dylib";
@@ -401,7 +401,7 @@
 
 #if TARGET_WINDOS
     global.params.defaultlibname = "phobos";
-#elif TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS
+#elif TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_HITOMI
     global.params.defaultlibname = "phobos2";
 #elif TARGET_NET
 #else
@@ -441,6 +441,10 @@
     VersionCondition::addPredefinedGlobalIdent("Posix");
     VersionCondition::addPredefinedGlobalIdent("Solaris");
     global.params.isSolaris = 1;
+#elif TARGET_HITOMI
+    VersionCondition::addPredefinedGlobalIdent("Posix");
+    VersionCondition::addPredefinedGlobalIdent("Hitomi");
+    global.params.isHitomi = 1;
 #else
 #error "fix this"
 #endif
@@ -479,7 +483,7 @@
                 global.params.link = 0;
             else if (strcmp(p + 1, "cov") == 0)
                 global.params.cov = 1;
-#if TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS
+#if TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_HITOMI
             else if (strcmp(p + 1, "fPIC") == 0)
                 global.params.pic = 1;
 #endif
diff -ruN gdc-orig/d/dmd2/mars.h gdc/d/dmd2/mars.h
--- gdc-orig/d/dmd2/mars.h	2011-09-16 07:42:18.924516713 +0200
+++ gdc/d/dmd2/mars.h	2011-09-16 07:35:58.544511046 +0200
@@ -57,6 +57,7 @@
         TARGET_FREEBSD  Covers 32 and 64 bit FreeBSD
         TARGET_OPENBSD  Covers 32 and 64 bit OpenBSD
         TARGET_SOLARIS  Covers 32 and 64 bit Solaris
+        TARGET_HITOMI   Covers 32 and 64 bit Hitomi
         TARGET_NET      Covers .Net
 
     It is expected that the compiler for each platform will be able
@@ -105,7 +106,7 @@
 #define MODULEINFO_IS_STRUCT DMDV2   // if ModuleInfo is a struct rather than a class
 
 // Set if C++ mangling is done by the front end
-#define CPP_MANGLE (DMDV2 && (TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS))
+#define CPP_MANGLE (DMDV2 && (TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_HITOMI))
 
 /* Other targets are TARGET_LINUX, TARGET_OSX, TARGET_FREEBSD, TARGET_OPENBSD and
  * TARGET_SOLARIS, which are
@@ -121,7 +122,7 @@
 #endif
 #endif
 
-#if TARGET_LINUX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS
+#if TARGET_LINUX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_HITOMI
 #ifndef ELFOBJ
 #define ELFOBJ 1
 #endif
@@ -165,6 +166,7 @@
     char isFreeBSD;     // generate code for FreeBSD
     char isOPenBSD;     // generate code for OpenBSD
     char isSolaris;     // generate code for Solaris
+    char isHitomi;      // generate code for Hitomi
     char scheduler;     // which scheduler to use
     char useDeprecated; // allow use of deprecated features
     char useAssert;     // generate runtime code for assert()'s
diff -ruN gdc-orig/d/dmd2/mtype.c gdc/d/dmd2/mtype.c
--- gdc-orig/d/dmd2/mtype.c	2011-09-16 07:42:18.934516713 +0200
+++ gdc/d/dmd2/mtype.c	2011-09-16 07:35:58.584511046 +0200
@@ -87,7 +87,7 @@
 int REALSIZE = 16;
 int REALPAD = 6;
 int REALALIGNSIZE = 16;
-#elif TARGET_LINUX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_UNIX
+#elif TARGET_LINUX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_UNIX || TARGET_HITOMI
 int REALSIZE = 12;
 int REALPAD = 2;
 int REALALIGNSIZE = 4;
@@ -282,7 +282,7 @@
     if (global.params.is64bit)
     {
         PTRSIZE = 8;
-        if (global.params.isLinux || global.params.isFreeBSD || global.params.isSolaris)
+        if (global.params.isLinux || global.params.isFreeBSD || global.params.isSolaris || global.params.isHitomi)
         {
             REALSIZE = 16;
             REALPAD = 6;
@@ -297,7 +297,7 @@
 #if TARGET_OSX
         REALSIZE = 16;
         REALPAD = 6;
-#elif TARGET_LINUX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_UNIX
+#elif TARGET_LINUX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_UNIX || TARGET_HITOMI
         REALSIZE = 12;
         REALPAD = 2;
 #else
@@ -2526,7 +2526,7 @@
             sz = REALALIGNSIZE;
             break;
 
-#if TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_UNIX
+#if TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TARGET_SOLARIS || TARGET_UNIX || TARGET_HITOMI
         case Tint64:
         case Tuns64:
             sz = global.params.is64bit ? 8 : 4;
diff -ruN gdc-orig/d/target-ver-syms.sh gdc/d/target-ver-syms.sh
--- gdc-orig/d/target-ver-syms.sh	2011-09-16 07:42:19.984516730 +0200
+++ gdc/d/target-ver-syms.sh	2011-09-10 02:30:08.452486090 +0200
@@ -45,6 +45,7 @@
 sysv4*) d_os_versym=sysv4; d_unix=1 ;;
 *bsd*) d_os_versym=bsd; d_unix=1 ;;
 gnu*) d_os_versym=hurd; d_unix=1 ;;
+hitomi*) d_os_versym=hitomi; d_unix=1 ;;
 
 *) d_os_versym="$d_target_os"
 esac
@@ -90,5 +91,6 @@
 openbsd) echo "#define TARGET_OPENBSD 1" ;;
 solaris) echo "#define TARGET_SOLARIS 1" ;;
 Win32)   echo "#define TARGET_WINDOS  1" ;;
+hitomi)  echo "#define TARGET_HITOMI  1" ;;
 *)       echo "#define TARGET_UNIX    1" ;;
 esac
